### DISCUSSION TITLE
Crystal Plasticity Slow Computation

### DISCUSSION THREAD

**Discussion by smithht** [idaholab/moose#27931] [Q&A Modules: Solid mechanics]
Hi there,

I have been creating a dislocation density based crystal plasticity app in MOOSE. The app inherits from `ComputeMultipleCrystalPlasticityStress` and `CrystalPlasticityStressUpdateBase`. The app has been running just fine with smaller (~500 elements) simulations. As I have increased the size of my simulations to run more detailed microstructures, I have run into massive slowdowns. As an example, I ran a 36,000 element simulation on an HPC using 82 cores. It ran for 8 days, only getting to a strain of 0.00022 (0.022%) : 
<img width="615" alt="AFRL Challenge #4 Calibration" src="https://github.com/idaholab/moose/assets/150199080/891150bb-eb87-452a-a9d5-48295cf74d2a">
I need to get to a strain of at least 1%, and I would also like to increase the resolution of my microstructure even further (to ~300k elements). However, the speed of my app right now is far too slow to come close to these goals. Looking into other discussion posts about a slow computation time for a crystal plasticity simulation, I see that a simulation of this size should be much faster: 
> I see you are running with about 20k elements, for instance with 24 cpus, I would expect that simulation to take about 1 hour for each 1% strain load #26598 

I have profiled my code, and it revealed that most of the runtime was being spent calculating the Jacobian and the residual. A picture of the profile is found here: 
![run](https://github.com/idaholab/moose/assets/150199080/90760d29-7a62-4410-9929-cecdaf9f5274)
Much of the runtime is spent in `calculateConstitutiveSlipDerivative` , and specifically `pow` within that function. Here is a snippet of that function:
```
void
CrystalPlasticityDDStressUpdate::calculateConstitutiveSlipDerivative(std::vector<Real> & dslip_dtau)
{
  Real effective_stress;
  Real delF0 = (_enthalphy_const * _g * 1.0e6) * (std::pow(_burg_mag, 3.0)) * 1.0e-9;
  // Boltzmann constant
  Real B_k = 1.3806503e-23;
  for (const auto i : make_range(_number_slip_systems))
  {
    effective_stress = _tau[_qp][i] - _backstress[_qp][i];
    if (std::abs(effective_stress) <= _s_a[_qp][i])
      dslip_dtau[i] = 0.0;
    else if (MooseUtils::absoluteFuzzyEqual(_tau[_qp][i], 0.0))
      dslip_dtau[i] = 0.0;
    else
    {
      dslip_dtau[i] =
          _slip_rate_reference *
          std::exp((-delF0 / B_k / _temperature) *
                   std::pow(1.0 - std::pow((std::abs(effective_stress) - _s_a[_qp][i]) / _s_t,
                   _p),
                            _q)) *
          (delF0 / B_k / _temperature) * _q *
          std::pow(1.0 - std::pow((std::abs(effective_stress) - _s_a[_qp][i]) / _s_t, _p),
                   _q - 1.0) *
          _p * std::pow((std::abs(effective_stress) - _s_a[_qp][i]) / _s_t, _p - 1.0) / _s_t *
          sgn(effective_stress);
    }
    
  }
}
```
I do not see any obvious points in this function that would be causing major slowdowns. I have tried replacing the pow function with a faster rough approximation, but that did not help. 

I believe a next step for me might be to improve the preconditioning? But I am not sure where to start with that. What type of preconditioning might give me better results for large ( > 30k elements) simulations? How would that implementation look for a CP app like this? Otherwise, would hardcoding the Jacobian improve my results? I have read that the Jacobian is calculated as part of `ComputeMultipleCrystalPlasticityStress`, but would hardcoding it myself improve the results? Where and how would I implement that in my app? 

Here is an example of an input file I have been using: 
```
[GlobalParams]
  displacements = 'ux uy uz'
  use_displaced_mesh = true
[]

[Mesh]
  [cube]
    type = EBSDMeshGenerator
    filename = 'InitialStructure.txt'
    show_info = true
  []
  [bot_corner]
    type = ExtraNodesetGenerator
    new_boundary = bot_corner
    input = cube
    coord = '0.0 0.0 0.0'
  []
  [bot_x_corner]
    type = ExtraNodesetGenerator
    new_boundary = bot_x_corner
    input = bot_corner
    coord = '3.0 0.0 0.0'
  []
  [bot_z_corner]
    type = ExtraNodesetGenerator
    new_boundary = bot_z_corner
    input = bot_x_corner
    coord = '0.0 0.0 3.0'
  []
[]

[AuxVariables]
  [pk2]
    order = CONSTANT
    family = MONOMIAL
  []
  [fp_yy]
    order = CONSTANT
    family = MONOMIAL
  []
  [rotout]
    order = CONSTANT
    family = MONOMIAL
  []
  [e_yy]
    order = CONSTANT
    family = MONOMIAL
  []
  [gss]
    order = CONSTANT
    family = MONOMIAL
  []
  [slip_increment]
    order = CONSTANT
    family = MONOMIAL
  []
  [euler_angle_1]
    order = CONSTANT
    family = MONOMIAL
  []
  [euler_angle_2]
    order = CONSTANT
    family = MONOMIAL
  []
  [euler_angle_3]
    order = CONSTANT
    family = MONOMIAL
  []
  [s_a]
    order = CONSTANT
    family = MONOMIAL
  []
  [grain_id]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Modules]

  [TensorMechanics]

    [Master]

      [all]
        strain = FINITE
        add_variables = true
        generate_output = stress_yy
      []
    []
  []
[]

[AuxKernels]
  [fp_yy]
    type = RankTwoAux
    variable = fp_yy
    rank_two_tensor = plastic_deformation_gradient
    index_j = 1
    index_i = 1
    execute_on = timestep_end
  []
  [pk2]
    type = RankTwoAux
    variable = pk2
    rank_two_tensor = second_piola_kirchhoff_stress
    index_j = 1
    index_i = 1
    execute_on = timestep_end
  []
  [e_yy]
    type = RankTwoAux
    variable = e_yy
    rank_two_tensor = total_lagrangian_strain
    index_j = 1
    index_i = 1
    execute_on = timestep_end
  []
  [gss]
    type = MaterialStdVectorAux
    variable = gss
    property = slip_resistance
    index = 0
    execute_on = timestep_end
  []
  [slip_inc]
    type = MaterialStdVectorAux
    variable = slip_increment
    property = slip_increment
    index = 0
    execute_on = timestep_end
  []
  [euler_angle_1]
    type = MaterialRealVectorValueAux
    variable = euler_angle_1
    property = updated_Euler_angle
    component = 0
    execute_on = timestep_end
  []
  [euler_angle_2]
    type = MaterialRealVectorValueAux
    variable = euler_angle_2
    property = updated_Euler_angle
    component = 1
    execute_on = timestep_end
  []
  [euler_angle_3]
    type = MaterialRealVectorValueAux
    variable = euler_angle_3
    property = updated_Euler_angle
    component = 2
    execute_on = timestep_end
  []
  [s_a]
    type = MaterialStdVectorAux
    variable = s_a
    property = s_a
    index = 0
    execute_on = timestep_end
  []
  [grain_id]
    type = MaterialRealAux
    variable = grain_id
    property = grain_id
    execute_on = timestep_end
  []
[]

[BCs]
  [y_roller]
    type = DirichletBC
    preset = true
    variable = uy
    boundary = bottom
    value = 0.0
  []
  [x_corner]
    type = DirichletBC
    preset = true
    variable = uz
    boundary = bot_x_corner
    value = 0.0
  []
  [z_corner]
    type = DirichletBC
    preset = true
    variable = ux
    boundary = bot_z_corner
    value = 0.0
  []

  [z_bot]
    type = DirichletBC
    preset = true
    variable = ux
    boundary = bot_corner
    value = 0.0
  []

  [y_bot]
    type = DirichletBC
    preset = true
    variable = uy
    boundary = bot_corner
    value = 0.0
  []

  [x_bot]
    type = DirichletBC
    preset = true
    variable = uz
    boundary = bot_corner
    value = 0.0
  []

  [tdisp]
    type = FunctionDirichletBC
    variable = uy
    boundary = top
    function = '0.01*t'
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeElasticityTensorCPCustom
    C_ijkl = '243300.0 156700.0 156700.0 243300.0 156700.0 243300.0 111125.0 111125.0 111125.0'
    # C11 C12 C12 C11 C12 C11 C44 C44 C44
    fill_method = symmetric9
    Dream3DReader = Dream3DReaderObject
  []
  [stress]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact

    rtol = 1e-4 # Constitutive stress residual relative tolerance
    # maxiter_state_variable = 50 # Maximum number of iterations for stress update
    maximum_substep_iteration = 2 # Maximum number of substep iteration

    use_line_search = true
  []
  [trial_xtalpl]
    type = CrystalPlasticityDDStressUpdate
    #type = CrystalPlasticityKalidindiUpdate
    crystal_lattice_type = FCC
    number_slip_systems = 12
    slip_sys_file_name = fcc_slip_sys.txt
    Dream3DReader = Dream3DReaderObject

    g = 111125 # MPa
    burg_mag = 2.53e-07

    # params
    slip_rate_reference = 4000000.0
    enthalpy_const = 0.25
    p = 0.35
    q = 1.34140625
    p0 = 1.65625
    k_rho = 0.2845
    k_i = 0.12
    s_t = 290.0
    r = 0.7937500000000001
    h = 5.078125
    rho_m_initial = 500000.0
    rho_i_initial = 371093.75
    k_m = 0.0363281249999999
    r_c = 3.919375000000001e-06
    k_ann = 2.0
    k_d = 57.34375
    k_bs1 = 74.53125
    k_bs2 = 4812.5


    slip_increment_tolerance = 1.0 # Maximum allowable slip in an increment
    stol = 1.0 # Constitutive internal state variable relative change tolerance
    resistance_tol = 1.0
  []
  [updated_euler_angle]
    type = ComputeUpdatedEulerAngle
    radian_to_degree = true
  []
[]

[Postprocessors]
  [stress_yy]
    type = ElementAverageValue
    variable = stress_yy
  []
  [pk2]
    type = ElementAverageValue
    variable = pk2
  []
  [fp_yy]
    type = ElementAverageValue
    variable = fp_yy
  []
  [e_yy]
    type = ElementAverageValue
    variable = e_yy
  []
  [gss]
    type = ElementAverageValue
    variable = gss
  []
  [slip_increment]
    type = ElementAverageValue
    variable = slip_increment
  []

  [run_time]
    type = PerfGraphData
    section_name = "Root"
    data_type = total
  []

  [euler_angle_1]
    type = ElementAverageValue
    variable = euler_angle_1
  []
  [euler_angle_2]
    type = ElementAverageValue
    variable = euler_angle_2
  []
  [euler_angle_3]
    type = ElementAverageValue
    variable = euler_angle_3
  []

  [s_a]
    type = ElementAverageValue
    variable = s_a
  []
  [grain_id]
    type = ElementAverageValue
    variable = grain_id
  []
[]

[UserObjects]
  [Dream3DReaderObject]
    type = Dream3DDataReader
    file_name = 'InitialStructureASCII.txt'
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -ksp_type -ksp_gmres_restart'
  petsc_options_value = ' asm      2              lu            gmres     200'
  nl_abs_tol = 1e-10
  nl_rel_tol = 1e-10
  nl_abs_step_tol = 1e-10
  nl_max_its = 5 # Max number of nonlinear iterations

  start_time = 0.0
  end_time = 3.0
  #num_steps = 25
  dt = 0.005

  dtmin = 0.1e-9
  dtmax = 0.01
[]

[Outputs]
  file_base = 'out'
  csv = true
  exodus = true
  perf_graph = true # prints a performance report to the terminal
  #print_linear_residuals = true
  #console = false
  # [./console]
  #   type = Console
  #   output_linear = false
  #   max_rows = 5
  #   outlier_variable_norms = false
  #   enable = false
  #   #output_screen = false
  # [../]
[]

# [Debug]
#   show_var_residual_norms = true
# []
```

**Comment by GiudGiud** [comment]
Hello

is std::pow( used with an integer exponent by any chance?

**Reply by smithht** [reply]
No, they aren't integers. _p is ~0.35 and _q is ~1.3. So the exponents used in `calculateConstitutiveSlipDerivative` are 0.35, 1.3, -0.65, and 0.3 (or in that range. p and q can change by about +- 0.1)

**Reply by GiudGiud** [reply]
ok. We have an optimized pow() for integer exponents that's why I asked

for preconditioning,
```
  petsc_options_iname = '-pc_type'
  petsc_options_value = ' lu'
```
solves in very few iterations but could be slower still, especially for larger simulations

**Reply by smithht** [reply]
Hi, @GiudGiud , thank you for the suggestion. It looks like that petsc preconditioner helped improve the runtime by about 30%. However, the runtime for simulations with over 30,000 elements is still far too long to finish in a reasonable amount of time. Would changing the MOOSE preconditioner help in any way? What would changing that look like? Also, would hard coding the Jacobian yield any other improvements?

**Reply by GiudGiud** [reply]
> Also, would hard coding the Jacobian yield any other improvements?

This should be 30% faster than AD if you are up for the task. It may be even fasters with tensor computations.

> Would changing the MOOSE preconditioner help in any way?

Yes it can. However at 30k elements we don't expect any scaling issues. LU might already be one of the fastest options

**Reply by smithht** [reply]
Thanks, I will try to hand code the Jacobian then. I am a little confused where I need to implement this. Where is it that `computeQpJacobian()` needs to be overridden? This is something that belongs to a kernel, and not a material right? If I am using the SolidMechanics/Quasistatic master action:
```
[Physics/SolidMechanics/QuasiStatic/all]
  strain = FINITE
  add_variables = true
  generate_output = stress_yy
[]
```
how would I know what kernel to change? Thanks again for your help.

**Reply by GiudGiud** [reply]
If you use `Physics/SolidMechanics/QuasiStatic/all` then I dont think you need to implement anything
`use_automatic_differentiation` in this action defaults to false

**Reply by smithht** [reply]
Where do I override `computeQpJacobian()` ?

**Reply by GiudGiud** [reply]
in kernels

**Reply by smithht** [reply]
@GiudGiud , I think I may have misunderstood what you said here initially:
> If you use `Physics/SolidMechanics/QuasiStatic/all` then I dont think you need to implement anything `use_automatic_differentiation` in this action defaults to false

Were you saying that since the solid mechanics master action doesn’t use AD by default, that I won't gain anything from trying to hand code my Jacobian? 

On another note, @sapitts , I see that you have assisted with Crystal Plasticity questions in the past. Do you have any input as to why I might be having such bad scaling issues?

**Reply by GiudGiud** [reply]
If it does not use AD then it does benefit from implementing the Jacobian terms manually

**Reply by GiudGiud** [reply]
If you do use AD consistently everywhere, then no need for implementing Jacobians manually

**Reply by sapitts** [reply]
We have not implemented AD in the crystal plasticity codes because of the large number of stateful material properties (e.g. 6 material properties x 12 slip systems for even a simple FCC material). @dschwen and @dewenyushu may have more updates on current approaches to address this issue. There is also this capability added about a year ago to reduce the number of jacobian calculations: https://github.com/idaholab/moose/issues/23487
